compiler/src/iree/compiler/Dialect/HAL/Transforms/MaterializeDispatchInstrumentation.cpp

```diff
@@ -0,0 +1,400 @@
+static Value createPadding(Location loc, uint64_t unalignedLength,
+                           OpBuilder &builder) {
+  uint64_t padding = llvm::alignTo(unalignedLength, 16) - unalignedLength;
+  if (!padding) return nullptr;
+  auto i8Type = builder.getI8Type();
+  auto zeroAttr = IntegerAttr::get(i8Type, 0);
+  auto dataAttr = DenseElementsAttr::get(
+      VectorType::get({(int64_t)padding}, i8Type), zeroAttr);
+  return builder.create<IREE::Util::BufferConstantOp>(
+      loc, /*name=*/nullptr, dataAttr, builder.getIndexAttr(16),
+      /*mimeType=*/nullptr);
+}
+
+static void appendListItems(Location loc, Value list, ArrayRef<Value> items,
+                            OpBuilder &builder) {
+  Value oldLength = builder.create<IREE::Util::ListSizeOp>(loc, list);
+  Value newLength = builder.create<arith::AddIOp>(
+      loc, oldLength,
+      builder.create<arith::ConstantIndexOp>(loc, items.size()));
+  builder.create<IREE::Util::ListResizeOp>(loc, list, newLength);
+  for (size_t i = 0; i < items.size(); ++i) {
+    Value idx = builder.create<arith::AddIOp>(
+        loc, oldLength, builder.create<arith::ConstantIndexOp>(loc, i));
+    builder.create<IREE::Util::ListSetOp>(loc, list, idx, items[i]);
+  }
+}
+
+class MaterializeDispatchInstrumentationPass
+    : public PassWrapper<MaterializeDispatchInstrumentationPass,
+                         OperationPass<mlir::ModuleOp>> {
+ public:
+  MaterializeDispatchInstrumentationPass() = default;
+  MaterializeDispatchInstrumentationPass(
+      const MaterializeDispatchInstrumentationPass &pass) {}
+  explicit MaterializeDispatchInstrumentationPass(int64_t bufferSize) {
+    this->bufferSize = bufferSize;
+  }
+
+  StringRef getArgument() const override {
+    return "iree-hal-materialize-dispatch-instrumentation";
+  }
+
+  StringRef getDescription() const override {
+    return "Materializes dispatch instrumentation resources.";
+  }
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<mlir::arith::ArithDialect>();
+    registry.insert<IREE::HAL::HALDialect>();
+    registry.insert<IREE::Stream::StreamDialect>();
+    registry.insert<IREE::Util::UtilDialect>();
+  }
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    if (moduleOp.getBody()->empty()) return;
+
+    auto moduleBuilder = OpBuilder(&moduleOp.getBody()->front());
+    auto i8Type = moduleBuilder.getI8Type();
+    auto i32Type = moduleBuilder.getI32Type();
+    auto indexType = moduleBuilder.getIndexType();
+
+    // Used for all instrumentation.
+    auto loc = moduleBuilder.getUnknownLoc();
+
+    // Verify that the buffer size provided is a proper power of two.
+    if (!llvm::isPowerOf2_64(bufferSize)) {
+      llvm::errs() << "ERROR: iree-hal-materialize-dispatch-instrumentation "
+                      "pass requires a "
+                      "ringbuffer storage size that is a power of two; got "
+                   << std::to_string(bufferSize) << "\n";
```

Comment:
nit: could balance the linebreaks here
```suggestion
      llvm::errs() << "ERROR: iree-hal-materialize-dispatch-instrumentation "
                      "pass requires a ringbuffer storage size that is "
                      "a power of two; got "
                   << std::to_string(bufferSize) << "\n";
```

If there's a nice one-liner, you could also log the nearest (above or below) power of two, to make it easier for devs to copy/paste when tweaking their flag values.